provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinBuildScriptCompiler.kt

```diff
@@ -476,16 +476,44 @@ inline fun <T> KotlinScriptSource.withLocationAwareExceptionHandling(action: ()
 
 
 private
-inline fun <T> LoadedScriptClass<T>.eval(action: LoadedScriptClass<T>.() -> Unit) =
+inline fun <T> LoadedScriptClass<T>.eval(scriptSource: KotlinScriptSource, action: LoadedScriptClass<T>.() -> Unit) =
     withContextClassLoader(scriptClass.classLoader) {
-        try {
-            action()
-        } catch (e: InvocationTargetException) {
-            throw e.targetException
-        }
+        withLocationAwareScriptEvaluationExceptionHandling(scriptSource, action)
     }
 
 
+private
+inline fun <T> LoadedScriptClass<T>.withLocationAwareScriptEvaluationExceptionHandling(
+    scriptSource: KotlinScriptSource,
+    action: LoadedScriptClass<T>.() -> Unit
+) =
+    try {
+        action()
+    } catch (e: Throwable) {
+        throw maybeLocateException(scriptSource, e)
+    }
+
+
+private
+fun LoadedScriptClass<*>.maybeLocateException(scriptSource: KotlinScriptSource, e: Throwable): Throwable {
+    var actualException = e
+    if (actualException is InvocationTargetException) actualException = actualException.targetException
+
+    fun matches(element: StackTraceElement) =
+        element.className != null && (element.className == compiledScript.className || element.className.startsWith("${compiledScript.className}${'$'}"))
+
+    tailrec fun inferLocationFrom(exception: Throwable): Throwable {
+        if (exception is LocationAwareException) return exception
+        exception.stackTrace.find { matches(it) }
+            ?.run { return LocationAwareException(actualException, scriptSource.source, lineNumber.takeIf { it >= 0 }) }
```

Comment:
❓ Won't returning without throwing this exception not populate it's stack trace?

Reply:
Nope, stacktrace is populated in Exception constructors, see `java.lang.Throwable`.